File: NeetCode/Binary_Search/Find_Minimum_In_Rotated_Array/find__minimum__in__rotated__array.py
from typing import Optional, List


class Solution:
    def findMin(self, nums: List[int]) -> int:
        left = 0
        right = len(nums) - 1

        while left <= right:
            if nums[left] <= nums[right]:
                return nums[left]

            mid = (left + right) // 2

            if nums[mid] > nums[right]:
                left = mid + 1
            else:
                right = mid

        return nums[left]

        pass


if __name__ == "__main__":
    sol = Solution()
    nums = nums = [4, 5, 1, 2, 3]

    print(sol.findMin(nums))
    # print(sol.findMin([3, 4, 5, 1, 2]))
    print("Running Solution...")

File: NeetCode/Binary_Search/Find_Minimum_In_Rotated_Array/Find_Minimum_In_Rotated_Array.cpp
// Find_Minimum_In_Rotated_Array.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    int findMin(vector<int> &nums)
    {
        int n = nums.size();
        int left{0}, mid{0}, right = n - 1;

        while (left < right)
        {
            if (nums[left] < nums[right])
            {
                return nums[left];
            }

            mid = left + (right - left) / 2;

            if (nums[mid] > nums[right])
            {
                left = mid + 1;
            }
            else
            {
                right = mid;
            }
        }
        return nums[left];
    }
};
